executor/windowfuncs/window_funcs.go

```diff
@@ -0,0 +1,85 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package windowfuncs
+
+import (
+	"errors"
+
+	"github.com/pingcap/tidb/executor/aggfuncs"
+	"github.com/pingcap/tidb/expression/aggregation"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/util/chunk"
+)
+
+// WindowFunc is the interface for processing window functions.
+type WindowFunc interface {
```

Comment:
It seems for no frame window functions, we can directly use `aggfuncs.AggFunc` interface? Except that we need to specially record number of rows inside each partition?

And in the current window executor implementation, numPartitionRows can be hold in the window operator itself. Even if we implement a hash window operator, this information can still be hold in that operator.

So how about just use `AggFunc` interface nd remove the `remained` field in `aggWithoutFrame`?

Reply:
For no frame window functions, we can do it. But for other window functions like `row_number`, which is different from agg and can be handled more efficiently, the executor code may need code branches to handle it?

Reply:
Even if for no frame definition, we may need new interfaces. Currently, we will call `UpdatePartialResult` for all previous rows and call `AppendFinalResult2Chunk` after processing the group. For `row_number` function, this means we cannot directly append the result to chunk when updating partial results because the interface does not allow, or we need to use one more branch to handle the streaming function more efficiently.